### ISSUE TITLE
[Feature Request]: Support for zipped/tar dicom directories 

### ISSUE THREAD

**Issue by thomshaw92** [Donders-Institute/bidscoin#56]
[Feature Request] 
We have a bit of a problem keeping large numbers of DICOMS in our file servers before conversion. Is it possible to add support for reading tarballs of DICOMS into bidsmapper.py? 
Thanks!

**Comment by marcelzwiers** [comment]
Sure, I'll have a look. But the bidscoiner then needs to support it too I suppose?

**Comment by thomshaw92** [comment]
That would be ideal! Thank you! 
We get in trouble for all the inodes taken up by DICOMS and would ideally keep the DICOM dirs zipped. 
I think Heudiconv supports this by unzipping and re-zipping the directories as it goes.

**Comment by marcelzwiers** [comment]
Do you use / have a DICOMDIR file in the tarball?

**Comment by thomshaw92** [comment]
No I don't. Is this something I can generate? 

My raw data structure is usually
/sub-identifier/[ses-identifier]/seriesfolder/dicomfiles
Would ideally be
/sub-identifier/[ses-identifier]/seriesfolder.tar.gz (containing dicomfiles)

**Comment by marcelzwiers** [comment]
No, it is something that would make it more complicated :-). Seriesfolder.tar.gz is quite doable I suppose

**Comment by marcelzwiers** [comment]
Is it:
/sub-identifier/[ses-identifier]/seriesfolder1.tar.gz (e.g. containing the localizer scan)
/sub-identifier/[ses-identifier]/seriesfolder2.tar.gz (e.g. containing the T1 scan)
or:
/sub-identifier/[ses-identifier]/seriesfolders.tar.gz (e.g. containing all scans of the session)

**Comment by thomshaw92** [comment]
We have instances where both can happen, depending on the researcher!
 
But let's stick with option one for consistency with the rest of BidsCoin

/sub-identifier/[ses-identifier]/seriesfolder1.tar.gz (e.g. containing the localizer scan)
/sub-identifier/[ses-identifier]/seriesfolder2.tar.gz (e.g. containing the T1 scan)

**Comment by marcelzwiers** [comment]
Is it:
/sub-identifier/[ses-identifier]/seriesfolder1.tar.gz (e.g. containing the localizer scan)
/sub-identifier/[ses-identifier]/seriesfolder2.tar.gz (e.g. containing the T1 scan)
or:
/sub-identifier/[ses-identifier]/seriesfolders.tar.gz (e.g. containing all scans of the session)

**Comment by marcelzwiers** [comment]
I implemented a first version. You can test it by installing the latest/development version:
pip install git+https://github.com/Donders-Institute/bidscoin

**Comment by Cadaei-Yuvxvs** [comment]
Thank you for implementing this! I encountered two issues but I got there in the end.

I had to add a `parents=True` argument here: https://github.com/Donders-Institute/bidscoin/blob/638b7a20994b3317466a57034a193afae47fbba3/bidscoin/bids.py#L382

Another issue I encountered was — changing modalities in the bidseditor GUI wiped the attributes blank. I got around this by using bidsmap.yaml from an earlier run.

**Comment by marcelzwiers** [comment]
Ah thanks! But are you sure you have been using the latest version? Because the tempdir is made at line 363 and the previous line should already do the job (I will actually remove line 363 now):
``
        tempfolder   = Path(tempfile.mkdtemp())/subid/sesid
        tempfolder.mkdir()
``
As for changing modalities and wiping the attributes, well, that is inconvenient but actually intentional. The reason is that the attributes may change between modalities and the bookkeeping then becomes a nightmare.

**Comment by Cadaei-Yuvxvs** [comment]
Sorry for the confusion! I installed the GitHub version from mid last week, so I haven't had the chance to use the latest version.

**Comment by marcelzwiers** [comment]
That's ok, I apparently worked on the code after I reported here

**Comment by thomshaw92** [comment]
I'm having a bit of trouble with the latest version. 
Not sure if this deserves its own issue or if it is zip file specific. 

Behaviour: 
![bidscoin_1](https://user-images.githubusercontent.com/31721414/78359559-a2175a00-75f8-11ea-9969-c3244f25ad43.png)
![bidscoin_2](https://user-images.githubusercontent.com/31721414/78359564-a479b400-75f8-11ea-8941-954f50a5d9a4.png)

The dicoms are zipped - when running bidsmapper they all are identified as extra data, but when i correct them to be dwi, anat, etc, they lose all of the metadata and it defaults to <SeriesDescription> etc as you can see from the images.

Downloaded the latest version from git before starting.
Cheers :)

**Comment by marcelzwiers** [comment]
Ah I see, that makes sense because I clean-up the tmp-folder after I extracted the data. But when changing the modality it wants to read the header again from the source file. Didn't think of that. But this issue must have been there from the beginning...

**Comment by marcelzwiers** [comment]
Btw, that's a lot of dcm files you got there. Tip: try using the `-t bidsmap_dccn` option and see if it works for you. I will try to fix your issue in the meantime

**Comment by marcelzwiers** [comment]
The work-around is to change and comment-out line 275 from:
```
273            # Clean-up the temporary unpacked data
274            if unpacked:
275                shutil.rmtree(session)
```
to:
```
273            # Clean-up the temporary unpacked data
274            if unpacked:
275                pass # shutil.rmtree(session)
```

**Comment by marcelzwiers** [comment]
Another work-around would be:
Unzip the first subject(s) yourself, i.e. until there are no more new data samples found (so if you have the same protocol for all your subjects you only have to unzip the first subject)

**Comment by thomshaw92** [comment]
OK I went with that last option for the time being :) Thanks! 
3.0 is looking good!

**Comment by marcelzwiers** [comment]
I created this patch 4a5a6f07c824f6b4e12aec4b072d7b6996d43197 that should solve it...

**Comment by marcelzwiers** [comment]
There is still one downside, you can't run the `bidseditor` as a stand-alone tool when you use zipped source data (as the unzipped data is deleted after the bids-editing is done). Of course you can re-run the `bidsmapper`

**Comment by thomshaw92** [comment]
That makes sense. I hope I only need to run bidsmapper once unless the protocol changes throughout the study anyway. :) Thanks again!

**Comment by marcelzwiers** [comment]
Well, I should be more specific, you can run the `bidseditor` stand-alone and change the BIDS values / names etc, but you will run in the same issue if you change modality. Btw, you must have seen a lot of warnings about the data not being found :-)

**Comment by thomshaw92** [comment]
I've got another little bug now when running bidscoiner after running the bidsmapper with the zipped files.

2020-04-04 10:47:08 - bidscoin - INFO ------------------- Subject 1/24 -------------------
2020-04-04 10:47:09 - bidscoin - WARNING Cannot determine the dataformat of: /data/raw_data/sub-001/ses-01
Traceback (most recent call last):
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/bin/bidscoiner", line 8, in <module>
    sys.exit(main())
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bidscoiner.py", line 604, in main
    sesprefix    = args.sesprefix)
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bidscoiner.py", line 503, in bidscoiner
    if not bidsmap[dataformat]['session']:
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/ruamel/yaml/comments.py", line 753, in __getitem__
    return ordereddict.__getitem__(self, key)
KeyError: ''


`$ ls /data/raw_data/sub-001/ses-01`

0001_tune_up_shim.zip        
0002_standard_shim.zip                          
0015_wip_diffusion_511e_1p8_b2500_64dir_dfc_tracew.zip  
    
 etc...

**Comment by marcelzwiers** [comment]
Ok, the provenance issue keeps coming back, I will implement a better solution in the form of a little provenance store (containing data samples)

**Comment by marcelzwiers** [comment]
I implemented this provenance store (which may also be useful when moving bids-folders around). You can test it by installing the latest/development version:
pip install git+https://github.com/Donders-Institute/bidscoin

**Comment by thomshaw92** [comment]
Good idea! 
Unfortunately more bugs when testing - I re-ran bidsmapper after upgrading:

`2020-04-04 23:31:36 - bidscoin - WARNING Could not determine the dataformat of the source data
2020-04-04 23:31:45 - bidscoin - INFO -------------- FINISHED! -------------------
2020-04-04 23:31:45 - bidscoin - INFO
2020-04-04 23:31:45 - bidscoin - INFO The following BIDScoin errors and warnings were reported:
2020-04-04 22:32:25 - bidscoin - ERROR Could not find plugin: '/opt/bidscoin/plugins_dccn/presentation2events.py'
2020-04-04 22:33:30 - bidscoin - ERROR Could not find plugin: '/opt/bidscoin/plugins_dccn/presentation2events.py'

`

**Comment by marcelzwiers** [comment]
Ah, that's a silly little bug (the plugins were either not uploaded or I accidentally hardcoded a path somehow). I will fix it right away (should be quick)

**Comment by marcelzwiers** [comment]
Ah, I see you used the bidsmap_dccn template. Solution: either open the bids-editor and go to `Option`, then remove the dccn plugin that is causing the problem. If that doesn't work, open the bidsmap_dccn.yaml file with a text editor and delete the corresponding line

**Comment by marcelzwiers** [comment]
Fixed it (beeaa6b56e164e743fd803793ded5e3047b608a9), but was that `WARNING Could not determine the dataformat of the source data` doing there? That worries me...

**Comment by marcelzwiers** [comment]
If you confirm everything is now working I will release 3.0.2 (otherwise I keep on releasing ;-))
Thanks for testing!

**Comment by thomshaw92** [comment]
Just trying it out now!

**Comment by thomshaw92** [comment]
Still getting the same error for every participant when running coiner:
```2020-04-05 00:47:27 - bidscoin - WARNING Cannot determine the dataformat of: /data/raw_data/sub-001/ses-01
```

**Comment by marcelzwiers** [comment]
Did you delete the old bids-folder? Because the `bidsmapper` will otherwise re-use the `bidsmap.yaml`-file in there (without the provenance info).

**Comment by thomshaw92** [comment]
Will try now from scratch with bidsmapper... Can I still use the bidsmap_dccn template?

**Comment by marcelzwiers** [comment]
Yes, I saw that too, fixed it (f7eb8bc740a7b99b9f1394271e6d26341fff05af)

**Comment by thomshaw92** [comment]
So close!
One Warning while running Bidsmapper
```
2020-04-05 01:47:09 - bidscoin - WARNING Could not determine sub/ses-id information, /data/raw_data/sub-001/ses-02/dum.my should
```
and bidseditor wouldnt open after bidsmapper finished
```
2020-04-05 01:48:11 - bidscoin - INFO Writing bidsmap to: /data/bids/code/bidscoin/bidsmap.yaml
2020-04-05 01:48:29 - bidscoin - INFO Opening the bidseditor
Traceback (most recent call last):
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/bin/bidsmapper", line 8, in <module>
    sys.exit(main())
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bidsmapper.py", line 352, in main
    interactive  = args.interactive)
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bidsmapper.py", line 314, in bidsmapper
    gui.setupUi(mainwin, bidsfolder, bidsmapfile, bidsmap_new, copy.deepcopy(bidsmap_new), template, dataformat, subprefix=subprefix, sesprefix=sesprefix)
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bidseditor.py", line 201, in setupUi
    self.set_tab_bidsmap()
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bidseditor.py", line 707, in set_tab_bidsmap
    self.update_subses_and_samples(self.output_bidsmap)
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bidseditor.py", line 607, in update_subses_and_samples
    modality, run, '', self.subprefix, self.sesprefix)
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bids.py", line 1217, in get_bidsname
    subid, sesid = get_subid_sesid(Path(run['provenance']), subid, sesid, subprefix, sesprefix)
  File "/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/bids.py", line 1181, in get_subid_sesid
    subid = [part for part in sourcefile.parent.parts if part.startswith(subprefix)][-1]
IndexError: list index out of range
```

**Comment by marcelzwiers** [comment]
Are you sure you are using the latest version (including f7eb8bc740a7b99b9f1394271e6d26341fff05af)? Because that's the warning I fixed/removed...

p.s. Shouldn't you be catching sleep now? ;-)

**Comment by thomshaw92** [comment]
Hmm. Maybe I'm making a mistake here: 
pip install git+https://github.com/Donders-Institute/bidscoin@f7eb8bc740a7b99b9f1394271e6d26341fff05af

Same error as before....
Saturday night in quarantine! Sleep schedule is gone!

**Comment by marcelzwiers** [comment]
Looks ok to me, but as you see in the patch, I removed/changed that warning. So your install doesn't seem to update somehow

**Comment by marcelzwiers** [comment]
So the warning should go, it is causing the bidscoiner error. If the warning is gone, but the error in bidscoin isn't, that could be because you have to start with a clean bids folder

**Comment by thomshaw92** [comment]
OK! Upgraded it correctly...coiner is working again! 

Then Editor does not display all of the different series anymore:

![bidscoin_3](https://user-images.githubusercontent.com/31721414/78456441-01ec2e80-76e7-11ea-9e37-f585bbf47769.png)
 
and then coiner throws
```
2020-04-05 02:39:59 - bidscoin - INFO ------------------- Subject 1/1 -------------------
2020-04-05 02:39:59 - bidscoin - WARNING Cannot determine the dataformat of: /data/raw_data/sub-001/ses-01
```

**Comment by marcelzwiers** [comment]
Ah, that helps! A committed another patch that should help. But I don't see how you now get less series? Can you perhaps send me the logfile or show the complete log of the bidsmapper?

**Comment by thomshaw92** [comment]
[bidsmapper.log.txt](https://github.com/Donders-Institute/bidscoin/files/4432940/bidsmapper.log.txt)

[bidscoiner.errors.txt](https://github.com/Donders-Institute/bidscoin/files/4432938/bidscoiner.errors.txt)

[bidscoiner.log.txt](https://github.com/Donders-Institute/bidscoin/files/4432939/bidscoiner.log.txt)

**Comment by marcelzwiers** [comment]
Ok, could it be that for instance `STIMMRI20-Q1876/data/raw_data/sub-001/ses-01/0009_wip944_mp2rage_0.75iso_7t_uni_images.zip` and `wip_diffusion_511e_1p8_b1000_32dir_dfc_fa.zip` have a file/folder substructure that is suited for `dicomsort` and the one that are not appearing in the `bidsmapper` have a different file/folder structure? Because what the the `bidsmapper` does is, unzip, run `dicomsort` on the unzipped data and then take it from there.

You still get a `bidscoiner` dcm2niix error of which I cannot see the cause. I now committed a patch that should provide a more meaningful message (339c62b51045c72f83faa542265240760cbb7a60)

**Comment by thomshaw92** [comment]
I've done a bit of digging and ran bidsmapper on two different filesystems. I think there is something going on with the I/O of the unzipper function to /tmp - and I'm not sure what is going on with dicomsort...

Here are the logs of the two attempts with the same data and the corresponding bidseditors - you can see some series are not being picked up by the bidsmapper or inconsistent between them. 

![bidscoin_4](https://user-images.githubusercontent.com/31721414/78553377-7d1b2500-784c-11ea-8ac1-19cfe00de21b.png)
![bidscoin_5](https://user-images.githubusercontent.com/31721414/78553385-7ee4e880-784c-11ea-939a-d926f59e8413.png)
[bidsmapper_faster.log](https://github.com/Donders-Institute/bidscoin/files/4438019/bidsmapper_faster.log)
[bidsmapper_win.log](https://github.com/Donders-Institute/bidscoin/files/4438020/bidsmapper_win.log)



To answer your question:  The file structure is correct within the zips as well.



As for the bidscoiner error:
```
bidscoin - ERROR Failed to run:
module add dcm2niix;dcm2niix -b y -z y -i n -f "sub-001_ses-01_acq-WIPDiffusion511E1p8b250064dirDFCTRACEW_dir-COL_run-1_echo-1" -o "/data/fastertemp/uqtshaw/brainstim/bids_nu/sub-001/ses-01/extra_data" "/tmp/tmpryasr3xz/sub-001/ses-01/015-WIP_Diffusion_511E_1p8_b2500_64dir_DFC_TRACEW"
Errorcode 0:
/bin/sh: 1: module: not found
```

**Comment by marcelzwiers** [comment]
Thanks, that helped a lot! The funny thing is that yesterday I was reading about a beginner mistake that every python programmer should be aware of, namely that your function defaults should be immutable as they are evaluated only once. Now this may perhaps (hopefully) be the cause of this problem, as you can see in your log the same tmpdir (`tmp/tmp3d41lfw0`) is used twice, whereas it should have been random (I did put the randomization statement in the default argument). I created yet another patch...

**Comment by thomshaw92** [comment]
Ha! I have made this exact error before with /tmp! Should have looked for it. Thanks! :D 

Do you know what is going on with the dcm2niix error?

**Comment by marcelzwiers** [comment]
Oh, and the bidscoiner error is because you are not using a module system, but luckily have dcm2niix on your path already. See:

https://bidscoin.readthedocs.io/en/latest/options.html#dcm2niix

And:

# Dcm2niix installation

BIDScoin relies on dcm2niix to convert the source imaging data to nifti. Please download and install dcm2niix yourself according to the instructions. When done, make sure that the path to the dcm2niix binary / executable is set correctly in the BIDScoin Options in the [path_to_bidscoin]/heuristics/bidsmap_template.yaml file or in the Site specific / customized template file.

**Comment by thomshaw92** [comment]
Got it - will update this path now or just use module in the future. 

Testing the patch now. :)

**Comment by marcelzwiers** [comment]
Ooops, sorry Thom, another patch (the last one I think, really ;-))

**Comment by thomshaw92** [comment]
Cool - only one warning 
```
/data/home/uqtshaw/anaconda3/envs/python3.6/lib/python3.6/site-packages/bidscoin/dicomsort.py:109: UserWarning: File already exists, cannot safely rename /tmp/tmp5bxugfct/sub-001/ses-01/00000001.dcm -> /tmp/tmp5bxugfct/sub-001/ses-01/040-VC_eja_svs_mslaser_HS2/00000001.dcm
  warnings.warn(f"File already exists, cannot safely rename {dicomfile} -> {newfilename}")
```

**Comment by marcelzwiers** [comment]
Nice, so it finally works. You had a name collision that I fixed in the meantime ;-) (a9f67d03228c0114c672b312f9f34240b7496803).

I will now release 3.0.2, thanks for your feedback and sorry you had to go through my untested patches! :-)
